File: src/interfaces/freetdm.ts
import { EventEmitter } from 'events';
import { Logger } from '../utils/logger';
import { FreeTDMConfig, FreeTDMSpan, FreeTDMChannel } from '../types';

export interface FreeTDMEvent {
  type: 'call_incoming' | 'call_outgoing' | 'call_answered' | 'call_hangup' | 'dtmf' | 'alarm';
  spanId: number;
  channelId: number;
  data?: any;
  timestamp: Date;
}

export interface FreeTDMCallInfo {
  uniqueId: string;
  spanId: number;
  channelId: number;
  direction: 'inbound' | 'outbound';
  callerNumber?: string;
  calledNumber?: string;
  state: 'idle' | 'dialing' | 'ringing' | 'answered' | 'busy' | 'hangup';
  startTime: Date;
  answerTime?: Date;
  endTime?: Date;
}

export interface FreeTDMSpanStatus {
  spanId: number;
  name: string;
  type: string;
  state: 'up' | 'down' | 'alarm';
  totalChannels: number;
  activeChannels: number;
  availableChannels: number;
  alarms: string[];
  signaling: string;
}

export class FreeTDMInterface extends EventEmitter {
  private config: FreeTDMConfig;
  private logger: Logger;
  private isRunning = false;
  private spans: Map<number, FreeTDMSpan> = new Map();
  private activeCalls: Map<string, FreeTDMCallInfo> = new Map();
  private spanStatus: Map<number, FreeTDMSpanStatus> = new Map();
  private nativeBinding: any; // Native C++ binding to FreeTDM

  constructor(config: FreeTDMConfig, logger: Logger) {
    super();
    this.config = config;
    this.logger = logger;
    this.initializeSpans();
  }

  private initializeSpans(): void {
    for (const span of this.config.spans) {
      this.spans.set(span.spanId, span);
      this.spanStatus.set(span.spanId, {
        spanId: span.spanId,
        name: span.name,
        type: span.trunk_type,
        state: 'down',
        totalChannels: span.channels.length,
        activeChannels: 0,
        availableChannels: span.channels.filter(ch => ch.enabled).length,
        alarms: [],
        signaling: span.channels.find(ch => ch.signaling)?.signaling || 'none'
      });
    }
  }

  async start(): Promise<void> {
    if (this.isRunning) {
      throw new Error('FreeTDM interface already running');
    }

    if (!this.config.enabled) {
      this.logger.info('FreeTDM interface disabled in configuration');
      return;
    }

    try {
      // Load FreeTDM native binding
      this.loadNativeBinding();

      // Initialize FreeTDM library
      await this.initializeFreeTDM();

      // Configure spans
      await this.configureSpans();

      // Start event monitoring
      this.startEventMonitoring();

      this.isRunning = true;
      this.logger.info('FreeTDM interface started successfully');
      this.emit('started');
    } catch (error) {
      this.logger.error('Failed to start FreeTDM interface', error);
      throw error;
    }
  }

  async stop(): Promise<void> {
    if (!this.isRunning) {
      return;
    }

    try {
      // Stop all active calls
      this.hangupAllCalls();

      // Stop spans
      await this.stopSpans();

      // Shutdown FreeTDM
      await this.shutdownFreeTDM();

      this.isRunning = false;
      this.logger.info('FreeTDM interface stopped');
      this.emit('stopped');
    } catch (error) {
      this.logger.error('Error stopping FreeTDM interface', error);
      throw error;
    }
  }

  private loadNativeBinding(): void {
    try {
      // In a real implementation, this would load a native C++ module
      // that interfaces with the FreeTDM library
      // this.nativeBinding = require('./native/freetdm_binding');
      
      // For now, we'll simulate the binding
      this.nativeBinding = {
        init: () => Promise.resolve(),
        configure: () => Promise.resolve(),
        shutdown: () => Promise.resolve(),
        makeCall: () => Promise.resolve(),
        hangupCall: () => Promise.resolve(),
        sendDTMF: () => Promise.resolve(),
        getSpanStatus: () => ({}),
        getChannelStatus: () => ({}),
        on: () => {}
      };
      
      this.logger.info('FreeTDM native binding loaded');
    } catch (error) {
      throw new Error(`Failed to load FreeTDM native binding: ${error}`);
    }
  }

  private async initializeFreeTDM(): Promise<void> {
    return new Promise((resolve, reject) => {
      try {
        this.nativeBinding.init({
          configFile: this.config.configFile,
          logLevel: 'info'
        });

        // Set up event callbacks
        this.nativeBinding.on('call_incoming', (event: any) => {
          this.handleIncomingCall(event);
        });

        this.nativeBinding.on('call_answered', (event: any) => {
          this.handleCallAnswered(event);
        });

        this.nativeBinding.on('call_hangup', (event: any) => {
          this.handleCallHangup(event);
        });

        this.nativeBinding.on('dtmf', (event: any) => {
          this.handleDTMF(event);
        });

        this.nativeBinding.on('alarm', (event: any) => {
          this.handleAlarm(event);
        });

        resolve();
      } catch (error) {
        reject(error);
      }
    });
  }

  private async configureSpans(): Promise<void> {
    for (const [spanId, span] of this.spans) {
      try {
        await this.configureSpan(span);
        this.updateSpanStatus(spanId, 'up');
        this.logger.info(`Span ${spanId} (${span.name}) configured and started`);
      } catch (error) {
        this.updateSpanStatus(spanId, 'alarm');
        this.logger.error(`Failed to configure span ${spanId}`, error);
      }
    }
  }

  private async configureSpan(span: FreeTDMSpan): Promise<void> {
    const spanConfig = {
      spanId: span.spanId,
      name: span.name,
      trunkType: span.trunk_type,
      dChannel: span.d_channel,
      channels: span.channels.map(ch => ({
        id: ch.id,
        type: ch.type,
        enabled: ch.enabled,
        signaling: ch.signaling
      }))
    };

    return new Promise((resolve, reject) => {
      try {
        this.nativeBinding.configure(spanConfig);
        resolve();
      } catch (error) {
        reject(error);
      }
    });
  }

  private startEventMonitoring(): void {
    // Start periodic status monitoring
    setInterval(() => {
      this.updateSpanStatuses();
    }, 5000); // Update every 5 seconds
  }

  private updateSpanStatuses(): void {
    for (const spanId of this.spans.keys()) {
      try {
        const status = this.nativeBinding.getSpanStatus(spanId);
        if (status) {
          this.updateSpanStatusFromNative(spanId, status);
        }
      } catch (error) {
        this.logger.warn(`Failed to get status for span ${spanId}`, error);
      }
    }
  }

  private updateSpanStatusFromNative(spanId: number, nativeStatus: any): void {
    const currentStatus = this.spanStatus.get(spanId);
    if (!currentStatus) return;

    const updated = {
      ...currentStatus,
      state: nativeStatus.state || currentStatus.state,
      activeChannels: nativeStatus.activeChannels || 0,
      availableChannels: nativeStatus.availableChannels || currentStatus.totalChannels,
      alarms: nativeStatus.alarms || []
    };

    this.spanStatus.set(spanId, updated);
    this.emit('spanStatusUpdate', updated);
  }

  // Call management methods
  async makeCall(spanId: number, channelId: number, destination: string): Promise<string> {
    if (!this.isRunning) {
      throw new Error('FreeTDM interface not running');
    }

    const span = this.spans.get(spanId);
    if (!span) {
      throw new Error(`Span ${spanId} not found`);
    }

    const channel = span.channels.find(ch => ch.id === channelId);
    if (!channel || !channel.enabled) {
      throw new Error(`Channel ${channelId} not available on span ${spanId}`);
    }

    const uniqueId = this.generateUniqueCallId();
    const callInfo: FreeTDMCallInfo = {
      uniqueId,
      spanId,
      channelId,
      direction: 'outbound',
      calledNumber: destination,
      state: 'dialing',
      startTime: new Date()
    };

    this.activeCalls.set(uniqueId, callInfo);

    try {
      await this.nativeBinding.makeCall({
        spanId,
        channelId,
        destination,
        uniqueId
      });

      this.emit('callInitiated', callInfo);
      return uniqueId;
    } catch (error) {
      this.activeCalls.delete(uniqueId);
      throw error;
    }
  }

  async hangupCall(uniqueId: string): Promise<void> {
    const callInfo = this.activeCalls.get(uniqueId);
    if (!callInfo) {
      throw new Error(`Call ${uniqueId} not found`);
    }

    try {
      await this.nativeBinding.hangupCall({
        uniqueId,
        spanId: callInfo.spanId,
        channelId: callInfo.channelId
      });

      callInfo.state = 'hangup';
      callInfo.endTime = new Date();
      this.emit('callHangup', callInfo);
    } catch (error) {
      this.logger.error(`Failed to hangup call ${uniqueId}`, error);
      throw error;
    }
  }

  async sendDTMF(uniqueId: string, digits: string): Promise<void> {
    const callInfo = this.activeCalls.get(uniqueId);
    if (!callInfo) {
      throw new Error(`Call ${uniqueId} not found`);
    }

    try {
      await this.nativeBinding.sendDTMF({
        uniqueId,
        spanId: callInfo.spanId,
        channelId: callInfo.channelId,
        digits
      });

      this.emit('dtmfSent', { uniqueId, digits });
    } catch (error) {
      this.logger.error(`Failed to send DTMF for call ${uniqueId}`, error);
      throw error;
    }
  }

  // Event handlers
  private handleIncomingCall(event: any): void {
    const uniqueId = this.generateUniqueCallId();
    const callInfo: FreeTDMCallInfo = {
      uniqueId,
      spanId: event.spanId,
      channelId: event.channelId,
      direction: 'inbound',
      callerNumber: event.callerNumber,
      calledNumber: event.calledNumber,
      state: 'ringing',
      startTime: new Date()
    };

    this.activeCalls.set(uniqueId, callInfo);
    this.emit('incomingCall', callInfo);
  }

  private handleCallAnswered(event: any): void {
    const callInfo = this.findCallBySpanChannel(event.spanId, event.channelId);
    if (callInfo) {
      callInfo.state = 'answered';
      callInfo.answerTime = new Date();
      this.emit('callAnswered', callInfo);
    }
  }

  private handleCallHangup(event: any): void {
    const callInfo = this.findCallBySpanChannel(event.spanId, event.channelId);
    if (callInfo) {
      callInfo.state = 'hangup';
      callInfo.endTime = new Date();
      this.activeCalls.delete(callInfo.uniqueId);
      this.emit('callHangup', callInfo);
    }
  }

  private handleDTMF(event: any): void {
    const callInfo = this.findCallBySpanChannel(event.spanId, event.channelId);
    if (callInfo) {
      this.emit('dtmfReceived', {
        uniqueId: callInfo.uniqueId,
        digit: event.digit
      });
    }
  }

  private handleAlarm(event: any): void {
    const spanStatus = this.spanStatus.get(event.spanId);
    if (spanStatus) {
      spanStatus.alarms.push(event.alarmType);
      spanStatus.state = 'alarm';
      this.emit('alarm', {
        spanId: event.spanId,
        alarmType: event.alarmType,
        severity: event.severity,
        message: event.message
      });
    }
  }

  // Utility methods
  private findCallBySpanChannel(spanId: number, channelId: number): FreeTDMCallInfo | undefined {
    for (const callInfo of this.activeCalls.values()) {
      if (callInfo.spanId === spanId && callInfo.channelId === channelId) {
        return callInfo;
      }
    }
    return undefined;
  }

  private generateUniqueCallId(): string {
    return `FTDM-${Date.now()}-${Math.random().toString(36).substr(2, 9)}`;
  }

  private hangupAllCalls(): void {
    const callIds = Array.from(this.activeCalls.keys());
    for (const callId of callIds) {
      this.hangupCall(callId).catch(error => {
        this.logger.warn(`Failed to hangup call ${callId} during shutdown`, error);
      });
    }
  }

  private async stopSpans(): Promise<void> {
    for (const spanId of this.spans.keys()) {
      this.updateSpanStatus(spanId, 'down');
    }
  }

  private async shutdownFreeTDM(): Promise<void> {
    return new Promise((resolve) => {
      try {
        this.nativeBinding.shutdown();
        resolve();
      } catch (error) {
        this.logger.warn('Error during FreeTDM shutdown', error);
        resolve();
      }
    });
  }

  private updateSpanStatus(spanId: number, state: 'up' | 'down' | 'alarm'): void {
    const status = this.spanStatus.get(spanId);
    if (status) {
      status.state = state;
      this.emit('spanStatusUpdate', status);
    }
  }

  // Public API methods
  getSpanStatus(spanId: number): FreeTDMSpanStatus | undefined {
    return this.spanStatus.get(spanId);
  }

  getAllSpanStatuses(): FreeTDMSpanStatus[] {
    return Array.from(this.spanStatus.values());
  }

  getActiveCall(uniqueId: string): FreeTDMCallInfo | undefined {
    return this.activeCalls.get(uniqueId);
  }

  getAllActiveCalls(): FreeTDMCallInfo[] {
    return Array.from(this.activeCalls.values());
  }

  getCallsForSpan(spanId: number): FreeTDMCallInfo[] {
    return Array.from(this.activeCalls.values()).filter(call => call.spanId === spanId);
  }

  isRunning(): boolean {
    return this.isRunning;
  }

  getConfiguration(): FreeTDMConfig {
    return { ...this.config };
  }
}

File: src/interfaces/freetdm.rs
//! FreeTDM interface implementation

use std::collections::HashMap;
use std::path::Path;

use tokio::sync::mpsc;
use tracing::info;

use crate::config::{FreeTdmConfig, ChannelType, SignalingType, Layer1Type};
use crate::{Error, Result};

/// FreeTDM span status
#[derive(Debug, Clone)]
pub struct SpanStatus {
    pub span_id: u32,
    pub name: String,
    pub trunk_type: Layer1Type,
    pub is_up: bool,
    pub channels: Vec<ChannelInfo>,
    pub alarms: Vec<String>,
}

/// Channel information
#[derive(Debug, Clone)]
pub struct ChannelInfo {
    pub id: u8,
    pub channel_type: ChannelType,
    pub state: ChannelState,
    pub signaling: SignalingType,
    pub enabled: bool,
}

/// Channel states
#[derive(Debug, Clone, PartialEq)]
pub enum ChannelState {
    Idle,
    InUse,
    Blocked,
    OutOfService,
}

/// FreeTDM events
#[derive(Debug, Clone)]
pub enum FreeTdmEvent {
    IncomingCall {
        span_id: u32,
        channel_id: u8,
        calling_number: Option<String>,
        called_number: Option<String>,
    },
    CallAnswered {
        span_id: u32,
        channel_id: u8,
    },
    CallHangup {
        span_id: u32,
        channel_id: u8,
        cause: u16,
    },
    Alarm {
        span_id: u32,
        message: String,
        severity: AlarmSeverity,
    },
    SpanUp {
        span_id: u32,
    },
    SpanDown {
        span_id: u32,
    },
}

#[derive(Debug, Clone)]
pub enum AlarmSeverity {
    Info,
    Warning,
    Critical,
}

/// FreeTDM interface wrapper
pub struct FreeTdmInterface {
    config: FreeTdmConfig,
    spans: HashMap<u32, SpanStatus>,
    event_tx: mpsc::UnboundedSender<FreeTdmEvent>,
    event_rx: Option<mpsc::UnboundedReceiver<FreeTdmEvent>>,
    is_running: bool,
}

impl FreeTdmInterface {
    pub fn new(config: FreeTdmConfig) -> Result<Self> {
        let (event_tx, event_rx) = mpsc::unbounded_channel();
        
        let mut spans = HashMap::new();
        
        // Initialize spans from configuration
        for span_config in &config.spans {
            let channels: Vec<ChannelInfo> = span_config.channels
                .iter()
                .map(|ch| ChannelInfo {
                    id: ch.id,
                    channel_type: ch.channel_type.clone(),
                    state: ChannelState::Idle,
                    signaling: ch.signaling.clone(),
                    enabled: ch.enabled,
                })
                .collect();

            let span_status = SpanStatus {
                span_id: span_config.span_id,
                name: span_config.name.clone(),
                trunk_type: span_config.trunk_type.clone(),
                is_up: false,
                channels,
                alarms: Vec::new(),
            };
            
            spans.insert(span_config.span_id, span_status);
        }

        Ok(Self {
            config,
            spans,
            event_tx,
            event_rx: Some(event_rx),
            is_running: false,
        })
    }

    pub fn take_event_receiver(&mut self) -> Option<mpsc::UnboundedReceiver<FreeTdmEvent>> {
        self.event_rx.take()
    }

    pub async fn start(&mut self) -> Result<()> {
        if !self.config.enabled {
            info!("FreeTDM interface is disabled");
            return Ok(());
        }

        info!("Starting FreeTDM interface");

        // Validate configuration file exists
        if !Path::new(&self.config.config_file).exists() {
            return Err(Error::parse(format!(
                "FreeTDM config file not found: {}",
                self.config.config_file
            )));
        }

        // In a real implementation, this would:
        // 1. Load and parse the FreeTDM configuration file
        // 2. Initialize the FreeTDM library
        // 3. Configure spans and channels
        // 4. Start event monitoring threads
        // 5. Set up signal handlers

        // For now, we'll simulate successful startup
        info!("FreeTDM interface started (simulated)");
        
        // Simulate spans coming up
        let span_ids: Vec<u32> = self.spans.keys().copied().collect();
        for span_id in span_ids {
            self.set_span_status(span_id, true).await;
        }

        self.is_running = true;
        Ok(())
    }

    pub async fn stop(&mut self) -> Result<()> {
        if !self.is_running {
            return Ok(());
        }

        info!("Stopping FreeTDM interface");

        // In a real implementation, this would:
        // 1. Stop all ongoing calls
        // 2. Shutdown FreeTDM spans
        // 3. Clean up resources
        // 4. Unload FreeTDM library

        self.is_running = false;
        info!("FreeTDM interface stopped");
        Ok(())
    }

    async fn set_span_status(&mut self, span_id: u32, is_up: bool) {
        if let Some(span) = self.spans.get_mut(&span_id) {
            span.is_up = is_up;
            
            let event = if is_up {
                FreeTdmEvent::SpanUp { span_id }
            } else {
                FreeTdmEvent::SpanDown { span_id }
            };
            
            let _ = self.event_tx.send(event);
            info!("Span {} is {}", span_id, if is_up { "UP" } else { "DOWN" });
        }
    }

    pub fn get_span_status(&self, span_id: u32) -> Option<&SpanStatus> {
        self.spans.get(&span_id)
    }

    pub fn get_all_span_statuses(&self) -> Vec<SpanStatus> {
        self.spans.values().cloned().collect()
    }

    pub fn is_running(&self) -> bool {
        self.is_running
    }

    pub async fn place_call(
        &self,
        span_id: u32,
        channel_id: u8,
        called_number: &str,
    ) -> Result<()> {
        if !self.is_running {
            return Err(Error::invalid_state("FreeTDM interface not running"));
        }

        // Validate span and channel exist
        let span = self.spans.get(&span_id)
            .ok_or_else(|| Error::tdm(format!("Span {} not found", span_id)))?;

        let channel = span.channels.iter()
            .find(|ch| ch.id == channel_id)
            .ok_or_else(|| Error::tdm(format!("Channel {} not found on span {}", channel_id, span_id)))?;

        if channel.state != ChannelState::Idle {
            return Err(Error::tdm(format!("Channel {}/{} is not idle", span_id, channel_id)));
        }

        // In a real implementation, this would initiate an outbound call
        // through the FreeTDM library
        info!("Placing call on span {}, channel {} to {}", span_id, channel_id, called_number);

        Ok(())
    }

    pub async fn answer_call(&self, span_id: u32, channel_id: u8) -> Result<()> {
        if !self.is_running {
            return Err(Error::invalid_state("FreeTDM interface not running"));
        }

        info!("Answering call on span {}, channel {}", span_id, channel_id);

        // Send answer event
        let _ = self.event_tx.send(FreeTdmEvent::CallAnswered {
            span_id,
            channel_id,
        });

        Ok(())
    }

    pub async fn hangup_call(&self, span_id: u32, channel_id: u8, cause: u16) -> Result<()> {
        if !self.is_running {
            return Err(Error::invalid_state("FreeTDM interface not running"));
        }

        info!("Hanging up call on span {}, channel {} with cause {}", span_id, channel_id, cause);

        // Send hangup event
        let _ = self.event_tx.send(FreeTdmEvent::CallHangup {
            span_id,
            channel_id,
            cause,
        });

        Ok(())
    }

    pub fn get_channel_count(&self) -> u32 {
        self.spans.values()
            .map(|span| span.channels.len() as u32)
            .sum()
    }

    pub fn get_active_channel_count(&self) -> u32 {
        self.spans.values()
            .flat_map(|span| &span.channels)
            .filter(|ch| ch.state == ChannelState::InUse)
            .count() as u32
    }
}

// Note: In a real implementation, you would create FFI bindings to the FreeTDM C library
// This would involve:
// 1. Creating a freetdm-sys crate with bindgen
// 2. Implementing safe wrappers around the C API
// 3. Setting up proper callbacks for events
// 4. Managing memory safety between Rust and C

#[cfg(test)]
mod tests {
    use super::*;
    use crate::config::{FreeTdmChannel, ChannelType, SignalingType, Layer1Type};

    #[tokio::test]
    async fn test_freetdm_interface_creation() {
        let config = FreeTdmConfig {
            enabled: false,
            config_file: "/tmp/test.conf".to_string(),
            spans: vec![],
        };
        
        let interface = FreeTdmInterface::new(config);
        assert!(interface.is_ok());
    }

    #[tokio::test]
    async fn test_freetdm_disabled() {
        let config = FreeTdmConfig {
            enabled: false,
            config_file: "/tmp/test.conf".to_string(),
            spans: vec![],
        };
        
        let mut interface = FreeTdmInterface::new(config).unwrap();
        let result = interface.start().await;
        assert!(result.is_ok());
        assert!(!interface.is_running());
    }
}